### PR TITLE
Constraint refactoring

### PR DESCRIPTION
**Pull request by ksaito7**
This PR refactors constraint stuff related to planner, executor, storage and catalog.

**- Contributions**
1. Add a new catalog table for constraints: `pg_constraint`
1. Clean up data structures of constraints and its usage:
Table constraint (Primary key, Foreign key, Unique, Check) -> `Constraint` (`pg_constraint`)
Column constraint (Not null, Default) -> `Column` (`pg_attribute`)
Deleted class -> `ForeignKey`, `MultiConstraint`
1. Add set/add/drop constraint functions in `Catalog`.
1. Simplify checking of constraints in `DataTable` when inserting/updating.

**- Related Issues**
1. #1270

**- Remaining Issues**
1. Parser does not support multiple column UNIQUE constraint.
1. CHECK constraints doesn't work when inserting/updating.
1. CHECK constraints doesn't support multiple columns or complex formula.
1. DDL: ADD/SET/DROP constraints are not supported.

### CODE REVIEW

**Comment by tli2** [comment]
@ksaito7 It looks like it will conflict with #1414 like crazy. Do you mind if you rebase on that PR locally?

**Comment by tli2** [comment]
@apavlo Joy wrote the original code and nobody has touched that code in a long time. I don't know if anybody else understands it so can you take a look?

**Comment by ksaito7** [comment]
@til2 Sure, I will rebase locally, and fix all tabs.

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -370,102 +396,28 @@ ResultType Catalog::CreateTable(const std::string &database_name,
   pg_table->InsertTable(table_oid, table_name, schema_name,
                         database_object->GetDatabaseOid(),
                         table->GetDefaultLayout()->GetOid(), pool_.get(), txn);
+
+  // Insert column info into each catalog
   oid_t column_id = 0;
+  std::vector<oid_t> pkey_attrs, unique_attrs;
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply src/catalog/catalog.cpp]: Delete them.

File: src/catalog/catalog.cpp [hunk 2]
```diff
@@ -370,102 +396,28 @@ ResultType Catalog::CreateTable(const std::string &database_name,
...
+  std::shared_ptr<Constraint> pkey_constraint, unique_constraint;
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply src/catalog/catalog.cpp]: Delete them.

File: src/catalog/catalog.cpp [hunk 3]
```diff
@@ -614,12 +570,315 @@ std::shared_ptr<const storage::Layout> Catalog::CreateDefaultLayout(
...
+                                        const std::string &constraint_name,
+                                        concurrency::TransactionContext *txn) {
+  auto table_object = catalog_map_[database_oid]->GetTableCatalog()
+                                                ->GetTableObject(table_oid, txn);
+  auto schema = storage::StorageManager::GetInstance()
+                    ->GetTableWithOid(database_oid, table_oid)
+                    ->GetSchema();
+
+  // Create index
+  std::stringstream index_name(table_object->GetTableName().c_str());
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply src/catalog/catalog.cpp]: Delete it.

File: src/catalog/constraint_catalog.cpp [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+      CONSTRAINT_CATALOG_CON_PKEY_OID, ConstraintType::PRIMARY, "con_primary",
+      CONSTRAINT_CATALOG_OID, std::vector<oid_t>{ColumnId::CONSTRAINT_OID},
+      CONSTRAINT_CATALOG_PKEY_OID));
+
+  return constraint_catalog_schema;
+}
+
+/*@brief    Insert a constraint into the pg_constraint table
+ *          This targets PRIMARY KEY, FOREIGN KEY, UNIQUE or CHECK constraint
+ * @param   table_oid  oid of the table related to this constraint
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply src/catalog/constraint_catalog.cpp]: Delete redundant comments.

File: src/include/catalog/table_catalog.h [hunk 5]
```diff
@@ -137,6 +154,11 @@ class TableCatalogObject {
       layout_objects_;
   bool valid_layout_objects_;
 
+  // cache for *all* layout objects in the table
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply src/include/catalog/table_catalog.h]: Modify it. I don't know what this * means. Emphasis? This is just a copy of other comments for the caches.

File: src/common/internal_types.cpp [hunk 6]
```diff
@@ -2147,6 +2132,62 @@ std::ostream &operator<<(std::ostream &os, const ConstraintType &type) {
   return os;
 }
 
+//===--------------------------------------------------------------------===//
+// Foreign Key Action Type - String Utilities
+//===--------------------------------------------------------------------===//
+
+std::string FKConstrActionTypeToString(FKConstrActionType type) {
+  switch (type) {
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply src/common/internal_types.cpp]: Add it.

File: src/catalog/constraint_catalog.cpp [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+  auto table_object = pg_table->GetTableObject(table_oid, txn);
+  PELOTON_ASSERT(table_object && table_object->GetTableOid() == table_oid);
+
+  auto constraint_object =
+      table_object->GetConstraintObject(constraint_oid, true);
+  if (constraint_object != nullptr) return constraint_object;
+
+  // cache miss, get from pg_attribute
+  std::vector<oid_t> column_ids(all_column_ids);
+  oid_t index_offset = IndexId::PRIMARY_KEY;  // Index of table_oid
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply src/catalog/constraint_catalog.cpp]: Fix it.

File: test/catalog/catalog_test.cpp [hunk 8]
```diff
@@ -79,14 +79,29 @@ TEST_F(CatalogTests, CreatingTable) {
...
+                       std::move(table_schema_3), txn);
+
+  auto emp =
+      catalog->GetTableObject("emp_db", DEFAULT_SCHEMA_NAME, "emp_table", txn);
+  auto department = catalog->GetTableObject("emp_db", DEFAULT_SCHEMA_NAME,
+                                            "department_table", txn);
+  auto salary = catalog->GetTableObject("emp_db", DEFAULT_SCHEMA_NAME,
+                                        "salary_table", txn);
+
+  catalog->AddPrimaryKeyConstraint(emp->GetDatabaseOid(), emp->GetTableOid(),
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply test/catalog/catalog_test.cpp]: These are not adding, but fixing. These might be used in other test? Verifying for primary keys is done in ConstraintCatalogTest.

File: test/catalog/catalog_test.cpp [hunk 9]
```diff
@@ -492,5 +507,231 @@ TEST_F(CatalogTests, LayoutCatalogTest) {
...
+            ConstraintTypeToString(con_object->GetConstraintType()).c_str());
+        EXPECT_TRUE(false);
+    }
+  }
+  txn_manager.CommitTransaction(txn);
+
+  LOG_DEBUG("%s", con_table->GetSchema()->GetInfo().c_str());
+  LOG_DEBUG("Complete check for constraint table");
+
+  // Drop constraint
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply test/catalog/catalog_test.cpp]: Correct. Add verifying.

File: test/concurrency/testing_transaction_util.cpp [hunk 10]
```diff
@@ -187,6 +202,14 @@ storage::DataTable *TestingTransactionUtil::CreateTable(
 
   table->AddIndex(pkey_index);
 
+  // Create primary key constraint on the table
+  if (need_primary_index) {
```
[the remark this answers is not in the document]
Comment (ksaito7) [reply test/concurrency/testing_transaction_util.cpp]: Fix it to need_primary_key.

**Pull request merged**